File: 线性表/CycleLinkList.java

import java.util.Random;

/** 循环单链表  这个Demo经测试已完成   但是toString打印就会导致无限递归出错
 * 但是内部已经形成指针类似的环
 * Created by han on 2016/10/30.
 */
public class CycleLinkList {

    public static void main(String[] args) {
        LinkList linkList = new LinkList();
        linkList.insertNode("ce1",1);
        linkList.insertNode("ce01", 1);
        linkList.insertNode("ce2", 3);
        linkList.GetElem(2);
        linkList.insertNode("ce02", 3);
        linkList.insertNode("ce4", 5);
        linkList.insertNode("ce3", 6);

        System.out.println("当前链表长度为：" + linkList.length());
        linkList.GetElem(1);
        linkList.delNode(1);
        linkList.delNode(2);
        linkList.GetElem(1);
        linkList.GetElem(2);
        System.out.println("当前链表长度为：" + linkList.length());
        linkList.GetElem(2);
        linkList.GetElem(3);
        System.out.println("------------------------------");
        linkList.CreateListHead(5);
        System.out.println("------------------------------");
        linkList.CreateListTail(5);
        linkList.GetElem(12);
        System.out.println("--------------------------------");
        linkList.GetElem(1);
        linkList.GetElem(5);
        linkList.GetElem(9);
        linkList.GetElem(10);
        linkList.GetElem(11);
        linkList.GetElem(12);
        linkList.GetElem(13);
        linkList.GetElem(14);
        linkList.ClearList();
        linkList.iterator();
    }
}
class LinkList{
    private   class Node{
        //结点数据
        Object data;
        //指向下一个结点指针
        Node next;
        public Node(){	}
        public Node(Object data ){
            this.data=data;
        }

       /* @Override//如果你要打印的话就产生了无限递归  因为是环
               public String toString() {
            return "Node{" +
                    "data=" + data +
                    ", next=" + next +
                    '}';
        }*/
    }

    private    Node head=new Node();//头结点


    public LinkList() {
        head.next = this.head;
        this.head = head.next;
    }//头结点在空结点时next的值是自身

    /*获取指定元素*/
    public  boolean GetElem(int i){
        int j;
        Node p; /*声明一指针p*/
        if(head ==null || i>length() )return  false;
        p=head.next; /*让p指向链表L的第一个结点*/
        j=1;
        while (p!=head && j<i){/*p不为空且计算器j还没有等于i时，循环继续*/
            p=p.next;/*让p指向下一个结点*/
            ++j;
        }
        if(p==head||j>i)/*第i个结点不存在*/
            return false;
//        p.data;/*取第i个结点的数据*/·
        System.out.println("Get到的数据"+p.data);
        return true;
    }



    /*插入指定位置*/
    public  boolean insertNode(Object o,int index){

        int i=1;
        Node n=new Node(o);
        Node preNode = head;//从头结点开始
        Node curNode = head.next;//设置这个变量为了循环记忆使用
        while (true) {

            if (i == index) {
                //这个步骤就是为了在prenode与curnode中插入节点 下面同理
                n.next = curNode;
                preNode.next = n;
                return true;
            }
            i++;
            preNode = curNode;
            curNode = preNode.next;
        }
    }


    /*删除指定节点*/
    public boolean delNode(int index){
        if(index <1||index >length()){
            return false;
        }
        int i=1;//计数器
        Node preNode=head;
        Node curNode=preNode.next;//设置这个变量为了循环记忆使用
        while (curNode!=null) {//从头结点向下循环寻找
            if(i==index){
                preNode.next=curNode.next;
                return true;
            }
            i++;
            preNode=curNode;
            curNode=preNode.next;
        }
        return true;
    }


    /*返回链表长度*/
    public int length(){
        int length=0;//计数器
        if (head.next == head) return  0;
        Node p=head.next;
        while (p!=head) {
            length++;
            p=p.next;
        }
        return length;
    }


    /*随机生成n个元素的值 建立带表头节点的单链线性表
     始终让头结点在第一个位置 头插法 头结点为空
    */
    public void CreateListHead( int n) {
        ++n;//因为不能插入头结点（p[0]） 所以通过自增来减掉偏差
        Node [] p=new Node[n];//要插入的节点
        int i;
        p[0]=head;//带头结点的单链表
        Random random = new Random( );

        for (i = 1; i < n; i++) {
            int  randomint=random.nextInt(100);
            p[i] = new Node(randomint);
            System.out.println("插入顺序"+"i:["+i+"]"+"\t"+"CreateListHead:"+randomint);
            p[i].next = p[0].next;
            p[0].next = p[i];//插入到表头
        }
        System.out.println("p[0]:" + p[0]);
        for (int j = 1; j < n; j++) {
            System.out.println("p["+j+"]"+p[j]);
        }

        System.out.println("当前链表长度为："+length());
    }





    /*随机产生n个元素的值 建立起带表头结点的单链线性表  尾插法*/
    public  void  CreateListTail(int n){
        Node [] node=new Node[n];//要插入的资源
        Node p,r;
        int i;
        Random random = new Random( );
        r= new Node();//初始化
        Node prenode=head;
        Node curnode = head.next;
        for (int j = 0; j <length() ; j++) {//遍历到当前最后一个结点
            curnode=prenode.next;
            prenode = curnode;
        }
        for (i = 0; i <n ; i++) {
            int rondomInt = random.nextInt(100);
            p = node[i] = new Node(rondomInt);// 每次都生成一个新节点:node[i] 随机填充数据
            r.next = p;// 将尾结点指针定位到当前结点
            p.next = head;
            r = p;// 将当前结点定义为新节点  （只是r的位置被改变 值本身还在原来的node数组中 r代表了尾结点）
            curnode.next = r;//在这里形成环
            prenode = curnode.next;
            curnode = prenode;
            System.out.println("插入顺序" + "i:[" + i + "]" + "\t" + "CreateListTail:" + rondomInt);
        }


        for ( i = n-1; i >=0 ; i--) {
            System.out.println( "node[" + i + "]:"+"CreateListTail:" + node[i]  );
        }

        System.out.println("CreateListTail当前链表长度为："+length());

    }

    /*迭代当前链表所有元素*/
    public  void iterator(){
        Node iteratorNode=head.next;
        Node iteratorNode2=iteratorNode.next;
        for (int ii = 2; ii <length() ; ii++) {
            iteratorNode=iteratorNode2;
            iteratorNode2=iteratorNode.next;
            System.out.println(iteratorNode.data);
        }
    }



    /*初始条件：顺序线性表L已存在 操作结果 ：将L重置为空表*/
    public  void ClearList(){
        head.next=head;
        head=head.next;

    }


    /*与C语言的主要处理方式是一样的 这个方法就没有进行完善*/
    public void mergeLinkList(LinkList L,LinkList L2){
        LinkList rearA,rearB,p,q;

    }


}

File: 线性表/CycleLinkList.c
#include <stdio.h>
#define OK 1
#define ERROR 0
#define TRUE 1
#define FALSE 0
typedef int ElemType;/*ElemType根据实际情况而定 这里假设为int*/
/*线性表的单链表存储结构  --循环链表  个人所改 这本书内部只是大概介绍了一下 注意 我个人对C语言的语法忘的差不多了
推荐看注释
或者看java版本来进行理解
只有一个Demo 在mergeLinkList
 需要注意 对错我是没有试验过的 bug应该有 只是通过了编译
 例行上传
*/
typedef int Status;
typedef struct Node {
	ElemType data;
	struct Node *next;
} Node;

typedef struct Node *LinkList;/*定义LinkList*/



//创建单向循环链表  初始化时，头结点的指针指向自己 注意：这个方法是正确的 这是经过测试的
LinkList CreateSgCcLinkList(void)
{
   int i,length=0,data=0;
   LinkList pTail=NULL,p_new=NULL;
   LinkList pHead=(LinkList)malloc(sizeof(Node));

    if(NULL == pHead){
    	printf("内存分配失败!\n");
    	exit(0);
	}
	pHead->data=0;
	pHead->next=pHead;//空节点状态 头结点指向自己
	pTail=pHead;//同理 空结点也是头结点
	 printf("请输入要创建链表的长度:");
	 scanf("%d",&length);

	 for(i=1;i<length+1;i++)
	 {
	 p_new=(LinkList)malloc(sizeof(Node));
	  if(NULL ==p_new){
	  	printf("内存分配失败! \n");
	  	exit(0);
	  }
	  printf("请输入第%d个结点的元素值："+i);
	  scanf("%d",&data);

	  p_new->data=data;
	  p_new->next=pHead;//最后一个结点总是指向头结点
	  pTail->next=p_new;//尾插法
	  pTail=p_new;
	  }

	  return pHead;

}

/*
初始条件：顺序线性表L已存在 1<=i<=ListLength(L)
操作条件：用e返回L中第i个数据元素的值
*/
Status GetElem(LinkList L,int i,ElemType *e) {
	int j;
	LinkList p,r; /*声明一指针p*/
	r=L;/*头结点 0结点*/
	p=L->next; /*让p指向链表L的第一个结点*/
	j=1; /*J为计数器*/
	while(p!=r && j<i) { /*p不为空且计算器j还没有等于i时，循环继续*/
		p=p->next;/*让p指向下一个结点*/
		++j;
	}
	if(p=r|| j>i)
		return ERROR;/*第i个结点不存在*/
	*e=p->data;/*取第i个结点的数据*/
	return OK;
}

/*
初始条件 ：顺序线性表L已存在 1<=i<=ListLength(L)
操作条件：在L中第i个结点位置之前插入新的数据元素e,l的长度加1
*/
Status ListInsert(LinkList *L,int i,ElemType e) {
	int j;
	LinkList p,s,r;
	r= *L;//头结点
	p=r->next;
	j=1;
	while(p!=r && j<i) { /*寻找第i-1个结点*/
		p=p->next;
		++j;
	}

	if(p=r || j>i)
		return ERROR;

	s=(LinkList)malloc(sizeof(Node));/* 生成新节点(C标准函数) sizeof操作符以字节形式给出了其操作数的存储大小*/
	s->data=e;
	s->next=p->next; /*将p的后继节点赋值给s的后继*/
	p->next=s;/*将S赋值给p的后继*/
	return OK;

}
/*
 初始条件：顺序线性表L已存在，1<=i<=ListLength(L)
 操作结果: 删除L的第i个结点 并用E返回其值 L的长度减1
 由于是循环链表  所以是不能等于头结点
*/
Status ListDelete(LinkList *L,int i,ElemType *e) {
	int j;
	LinkList p,q,r;
	r=*L;//头结点
	p=r->next;
	j=1;

	while(p->next!=r && j<i) { /*遍历寻找第i-1个结点*/
		p=p->next;
		++j;
	}

	if((p->next)=r || j>i)
		return ERROR;/*第i个结点不存在*/
	q=p->next;
	p->next=q->next;/*将q的后继赋值给p的后继*/
	*e=q->next;  /*将q结点中的数据给e**/
	free(q); /*让系统回收此节点 释放内存*/
	return OK;
}



/*随机生成n个元素的值 建立带表头节点的单链线性表*/
Status CreateListHead (LinkList *L,int n) {
	LinkList p,phead;
	int i;
	srand(time(0));/*初始化随机数种子*/
	*L=(LinkList)malloc(sizeof(Node));
	phead=(*L);/*先建立一个带头结点的单链表 这里理解（*L）是头结点*/
  phead->next=phead;/*空结点时 头结点本身是头结点*/

	for(i=0; i<n; i++) {

		p=(LinkList)malloc(sizeof(Node));/*生成新节点*/
		p->data=rand()%100+1;/*随机生成100以内的数字*/
		p->next=(*L)->next;
		(*L)->next=p;/*插入到表头*/

	}

}

/*循环链表特殊功能 合并俩个不同的链表*/
void mergeLinkList(LinkList *L,LinkList *L2){
LinkList rearA,rearB,p,q;
   rearA=*L;/*获取L尾结点   注意:我也不知道这样是否正确获取到*/
   rearB=*L2;/*获取L2尾结点 同上*/
   p=rearA->next;/*保存A表的头结点*/
   rearA->next=rearB->next->next; /*将本是指向B表的第一个结点（不是头结点）赋值给rearA->next*/
   q=rearB->next;
   rearB->next=p;
   free(q);/*释放q*/
}





/*随机产生n个元素的值 建立起带表头节点的单链线性表（尾插法）*/
void CreateListTail(LinkList *L,int n) {
	LinkList p,r,s;
	int i;
	srand(time(0));/*初始化随机数种子*/
	*L=(LinkList)malloc(sizeof(Node));/*为整个线性表*/
	r=*L;/*r此时为指向尾部的节点*/

	for(i=0; i<n; i++) {
		p=(Node *)malloc(sizeof(Node));/*生成新节点*/
		p->data=rand() %100+1;/*随机生成100以内的数字*/
		r->next=p;/*将表尾终端节点的指针指向新节点*/
		p->next=(*L);  //尾结点的下一个节点指向头结点
		r=p;//成为新的尾结点
		s=*L;//当前最后一个结点
		s->next=r;// 形成环
	}
	r->next=(*L);/*表示当前链表结束 尾结点的下一个是头结点*/

}



/*初始条件：顺序线性表L已存在 操作结果 ：将L重置为空表*/
Status ClearList(LinkList *L) {
	LinkList p,q,r;
	r=(*L);/*头结点*/
	p=(*L)->next; /*p指向第一个结点*/

	while(p!=r) {  /*没到表尾*/
		q=p->next;
		free(p);
		p=q;
	}
	(*L)->next=r;/*头结点指针域为头结点本身*/
	return OK;
}


//个人所加 并没有测试功能
int length(LinkList *L){
	int length=0;
	LinkList p=(*L);//获得头结点
	LinkList q=p->next;
	while(q!=p){
		length++;
		q=q->next;
	}
	return length;
}


void  main() {

	LinkList list,start;
	printf("请输入单链表的数据");
	list=CreateListHead(list,10);
	for(start =list->next; start!=NULL; start=start->data)
		printf("%d",start->data);
	print("/n");



}
